File: invocation-patterns/javascript/dead_code.js
// Copyright © 2026 Sthenos Security. All rights reserved.
// ============================================================================
// INVOCATION PATTERNS TEST — Case 3: Dead Code (JavaScript)
//
// Functions are NEVER called — not from routes, not from timers, not from
// any trigger. The module is NOT required by any entrypoint.
//
// Expected:
//   app_reachability = NOT_REACHABLE
//   taint_verdict    = N/A
// ============================================================================
'use strict';
const { execSync } = require('child_process');
const fs = require('fs');
const path = require('path');


function deadSqli(db, userInput) {
    // CWE-89: NOT_REACHABLE — never called, module never loaded
    return db.query(`SELECT * FROM users WHERE name = '${userInput}'`);
}

function deadCmdi(cmd) {
    // CWE-78: NOT_REACHABLE — never called
    return execSync(cmd).toString();
}

function deadPathTraversal(filename) {
    // CWE-22: NOT_REACHABLE — never called
    return fs.readFileSync(path.join('/var/data', filename), 'utf8');
}

function deadSsrf(url) {
    // CWE-918: NOT_REACHABLE — never called
    const http = require('http');
    return new Promise((resolve) => {
        http.get(url, (res) => {
            let data = '';
            res.on('data', chunk => data += chunk);
            res.on('end', () => resolve(data));
        });
    });
}

function deadEval(code) {
    // CWE-94: NOT_REACHABLE — never called
    return eval(code);
}

// This file has NO:
// - Express routes
// - setInterval/setTimeout
// - process.on handlers
// - IIFE or module-level calls
// - Class instantiation
// Pure dead code.

module.exports = { deadSqli, deadCmdi, deadPathTraversal, deadSsrf, deadEval };

File: invocation-patterns/python/dead_code.py
# Copyright © 2026 Sthenos Security. All rights reserved.
# ============================================================================
# INVOCATION PATTERNS TEST — Case 3: Dead Code
#
# These functions are NEVER called — not from HTTP routes, not from threads,
# not from init, not from anywhere. The module itself is NOT imported by any
# entrypoint.
#
# Expected:
#   app_reachability = NOT_REACHABLE (call graph finds no path)
#   taint_verdict    = N/A (analyzer skips NOT_REACHABLE findings)
# ============================================================================
"""Case 3: Dead code — functions that never execute."""
import os
import sqlite3
import subprocess


def dead_sqli(user_input: str):
    """CWE-89: SQL injection — but NEVER called.
    NOT_REACHABLE: no import path, no internal trigger.
    """
    conn = sqlite3.connect(':memory:')
    conn.execute(f"SELECT * FROM users WHERE name = '{user_input}'")
    conn.close()


def dead_cmdi(cmd: str):
    """CWE-78: Command injection — but NEVER called.
    NOT_REACHABLE: no import path, no internal trigger.
    """
    result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
    return result.stdout


def dead_path_traversal(filename: str):
    """CWE-22: Path traversal — but NEVER called.
    NOT_REACHABLE: no import path, no internal trigger.
    """
    path = os.path.join('/var/data', filename)
    with open(path) as f:
        return f.read()


def dead_ssrf(url: str):
    """CWE-918: SSRF — but NEVER called.
    NOT_REACHABLE: no import path, no internal trigger.
    """
    import urllib.request
    return urllib.request.urlopen(url).read()


def dead_eval(code: str):
    """CWE-94: Code injection — but NEVER called.
    NOT_REACHABLE: no import path, no internal trigger.
    """
    return eval(code)


# This file has NO:
# - Flask/FastAPI routes
# - threading/Timer calls
# - atexit/signal hooks
# - module-level function calls
# - class instantiation
# It is pure dead code.
